### PR TITLE
add tuning op support for ROCM ep

### PR DESCRIPTION
**Pull request by guotuofeng**
## Describe your changes
allow `tunable_op_enable` and `tunable_op_tuning_enable` for ROCM ep to speed up the performance.

The two flags need to be set to True for ROCMExecutionProvider to speed up the performance when running in AMD hardware.
This PR enable the support of ROCM ep speed up in perf tuning.
This PR also fix a bug when both metrics and model has inference settings by merging them. The metric.inference_settings has higher priority.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

### CODE REVIEW

File: olive/passes/onnx/perf_tuning.py [hunk 1]
```diff
@@ -171,6 +180,18 @@ def tune_onnx_model(perf_tuning_pass_ep, model, data_root, config):
     return optimized_model
 
 
+def rocm_tuning_enable(provider, provider_options):
```
Comment (trajepl) [inline olive/passes/onnx/perf_tuning.py:183]: to `enable_rocm_runing`？

File: olive/passes/onnx/perf_tuning.py [hunk 2]
```diff
@@ -163,6 +171,7 @@ def tune_onnx_model(perf_tuning_pass_ep, model, data_root, config):
         "execution_provider": best_result["execution_provider"],
         "provider_options": best_result["provider_options"],
         "io_bind": best_result["io_bind"],
+        "tuning_result": best_result.get("tuning_result"),
```
Comment (trajepl) [inline olive/passes/onnx/perf_tuning.py:174]: The `tuning_result` here stores the tuning op results, right? If so, how about to use `tuning_op_results`? It looks a bit confused with the variabel perf tuning's `tuning_results`.

File: olive/passes/onnx/perf_tuning.py [hunk 3]
```diff
@@ -171,6 +180,18 @@ def tune_onnx_model(perf_tuning_pass_ep, model, data_root, config):
     return optimized_model
 
 
+def rocm_tuning_enable(provider, provider_options):
+    assert provider == "ROCMExecutionProvider", "provider should be ROCMExecutionProvider"
+    # Please refer to the following links for the config or ROCMExecutionProvider
+    # https://github.com/microsoft/onnxruntime/blob/71657d1eb8b0a24a4b6584d9e904506a0b4e1521/onnxruntime/core/providers/rocm/rocm_execution_provider_info.cc#L24C1-L25
+    provider_options["tunable_op_enable"] = True
```
Comment (trajepl) [inline olive/passes/onnx/perf_tuning.py:187]: Just double check, following the link you attached, the accept value type is `char`. Does the ort python API make the transformation happen?
```c++
constexpr const char* kTunableOpEnable = "tunable_op_enable";
constexpr const char* kTunableOpTuningEnable = "tunable_op_tuning_enable";
```

File: olive/passes/onnx/perf_tuning.py (hunk 1 above)
Comment (guotuofeng) [reply olive/passes/onnx/perf_tuning.py:183]: will change to enable_rocm_op_tuning

File: olive/passes/onnx/perf_tuning.py (hunk 3 above)
Comment (guotuofeng) [reply olive/passes/onnx/perf_tuning.py:187]: the bool will be converted to 1/0 in c++ level.

**Review by trajepl — DISMISSED** [review]

File: olive/evaluator/olive_evaluator.py [hunk 4]
```diff
@@ -498,10 +500,9 @@ def _inference(
         else:
             logits = {k: torch.cat(logits[k], dim=0) for k in output_names}
 
+        tuning_result_file = inference_settings.get("tuning_result_file")
```
Comment (trajepl) [inline olive/evaluator/olive_evaluator.py:503]: should be "tuning_op_result"?

File: test/unit_test/model/test_onnx_model.py [hunk 5]
```diff
@@ -39,6 +39,161 @@ def test_model_prepare_session(mock_get_available_providers, inference_session_m
...
+                "provider_options": [{"device_id": "5"}],
+            },
+        ),
+    ],
+)
+@patch("onnxruntime.InferenceSession")
+@patch("onnxruntime.get_available_providers")
+def test_model_prepare_session_multiple_inference_settings(
+    mock_get_available_providers,
+    inference_session_mock,
```
Comment (trajepl) [inline test/unit_test/model/test_onnx_model.py:166]: Better to align the name mock_* or *_mock.

**mock_**get_available_providers,
inference_session**_mock**,

File: olive/model/handler/onnx.py [hunk 6]
```diff
@@ -77,9 +76,12 @@ def prepare_session(
...
-        inference_settings = deepcopy(inference_settings)
+        inference_settings_merged = {}
+        if self.inference_settings:
+            inference_settings_merged.update(self.inference_settings)
+        if inference_settings:
+            inference_settings_merged.update(inference_settings)
+        inference_settings = inference_settings_merged
 
         # if user doesn't not provide ep list, use default value([ep]). Otherwise, use the user's ep list
         # user provided ep list > eps given by arguments > default eps
```
Comment (trajepl) [inline olive/model/handler/onnx.py:87]: nit: Update the comment to align the variable name to make it clear.

**Review by trajepl — APPROVED** [review]
